test/jdk/com/sun/crypto/provider/Cipher/ChaCha20/UpdateAADTest.java

```diff
@@ -0,0 +1,129 @@
+        } catch (GeneralSecurityException gse) {
+            throw new RuntimeException("Failed to init static JCE components",
+                    gse);
+        }
+    }
+
+    public static void main(final String[] args) throws Exception {
+        ByteBuffer twoKBuf = ByteBuffer.allocate(2048);
+        ByteBuffer nonBABuf = ByteBuffer.allocate(1329);
+
+        System.out.println("----- Test 1: Baseline test -----");
+        System.out.println("Make an array backed buffer that is 16-byte " +
+                           "aligned, treat all data as AAD and feed it to " +
+                           " updateAAD.");
+        new AADUpdateTest(twoKBuf, true).run();
+
+        System.out.println("----- Test 2: Non Block Aligned Offset -----");
+        System.out.println("Use the same buffer, but place the offset such " +
+                           "that the remaining data is not block aligned.");
+        new AADUpdateTest(twoKBuf.position(395), true).run();
+
+        System.out.println("----- Test 3: Non Block Aligned Buf/Off -----");
+        System.out.println("Make a buffer of non-block aligned size with an " +
+                           "offset that keeps the remaining data non-block " +
+                           "aligned.");
+        new AADUpdateTest(nonBABuf.position(602), true).run();
+
+        System.out.println("----- Test 4: Aligned Buffer Slice -----");
+        System.out.println("Use a buffer of block aligned size, but slice " +
+                           "the buffer such that the slice offset is part " +
+                           "way into the original buffer.");
+        new AADUpdateTest(twoKBuf.rewind().slice(1024,1024).position(42),
+                true).run();
+
+        System.out.println("----- Test 5: Non-Aligned Buffer Slice -----");
+        System.out.println("Try the same test as #4, this time with " +
+                           "non-block aligned buffers/slices.");
+        new AADUpdateTest(nonBABuf.rewind().slice(347, 347).position(86),
+                true).run();
+    }
+
+    public static class AADUpdateTest implements Runnable {
+        private final ByteBuffer buffer;
+        private final boolean expectedPass;
+
+        AADUpdateTest(ByteBuffer buf, boolean expPass) {
+            buffer = Objects.requireNonNull(buf);
+            expectedPass = expPass;
+        }
+
+        @Override
+        public void run() {
+            Cipher cipher;
+            try {
+                SecretKey key = CC20GEN.generateKey();
+                byte[] nonce = new byte[12];
+                RAND.nextBytes(nonce);
+
+                cipher = Cipher.getInstance("ChaCha20-Poly1305");
+                cipher.init(Cipher.ENCRYPT_MODE, key,
+                        new IvParameterSpec(nonce));
+            } catch (GeneralSecurityException gse) {
+                throw new RuntimeException("Failed during test setup", gse);
+            }
+
+            try {
+                cipher.updateAAD(buffer);
```

Comment:
Shouldn't we also check that the cipher's state is what we expected after this update?

Reply:
I've added a check on the buffer position and limit to ensure that the buffer has no remaining data at the end.  This can only happen if processMultipleBlocks ends with the buffer's position left at the first byte of trailing data where 0 <= remaining < 16.  The engineUpdate method will take care of any remaining non-block aligned data (if it exists).